cpp/perspective/src/include/perspective/gnode.h

```diff
@@ -390,44 +378,56 @@ class PERSPECTIVE_EXPORT t_gnode {
-    const t_data_table& prev = *(m_oports[PSP_PORT_PREV]->get_table().get());
-    const t_data_table& current = *(m_oports[PSP_PORT_CURRENT]->get_table().get());
-    const t_data_table& transitions = *(m_oports[PSP_PORT_TRANSITIONS]->get_table().get());
-    const t_data_table& existed = *(m_oports[PSP_PORT_EXISTED]->get_table().get());
-    notify_context<CTX_T>(ctx, flattened, delta, prev, current, transitions, existed);
-}
+    
+    // Tables from the gnode which do not have the expressions applied yet
+    std::shared_ptr<t_data_table> delta = m_oports[PSP_PORT_DELTA]->get_table();
+    std::shared_ptr<t_data_table> prev = m_oports[PSP_PORT_PREV]->get_table();
+    std::shared_ptr<t_data_table> current = m_oports[PSP_PORT_CURRENT]->get_table();
+    std::shared_ptr<t_data_table> transitions = m_oports[PSP_PORT_TRANSITIONS]->get_table();
 
-/**
- * @brief Given multiple `t_data_table`s containing the different states of the context,
- * update the context with new data.
- *
- * Called on updates and additions AFTER a view is constructed from the table/context.
- *
- * @tparam CTX_T
- * @param ctx
- * @param flattened a `t_data_table` containing the flat data for the context
- * @param delta a `t_data_table` containing the changes to the dataset
- * @param prev a `t_data_table` containing the previous state
- * @param current a `t_data_table` containing the current state
- * @param transitions a `t_data_table` containing operations to transform the context
- * @param existed
- */
-template <typename CTX_T>
-void
-t_gnode::notify_context(CTX_T* ctx, const t_data_table& flattened, const t_data_table& delta,
-    const t_data_table& prev, const t_data_table& current, const t_data_table& transitions,
-    const t_data_table& existed) {
-    auto ctx_config = ctx->get_config();
+    // Existed is special - it will never have any expression columns and can
+    // be used as-is from the gnode.
+    const t_data_table& existed = *(m_oports[PSP_PORT_EXISTED]->get_table());
 
     ctx->step_begin();
-    // Flattened has the expressions at this point, as it has
-    // passed through the body of `process_table`.
-    ctx->notify(flattened, delta, prev, current, transitions, existed);
+
+    if (ctx->num_expressions() > 0) {
+        // Join expression tables on the context with gnode tables and pass those
+        // into the context so there is no distinction between expression and real
+        // columns for the context.
+        std::shared_ptr<t_expression_tables> ctx_expression_tables = ctx->get_expression_tables();
+
+        auto joined_flattened = flattened->join(ctx_expression_tables->m_flattened);
+        auto joined_delta = delta->join(ctx_expression_tables->m_delta);
+        auto joined_prev = prev->join(ctx_expression_tables->m_prev);
+        auto joined_current = current->join(ctx_expression_tables->m_current);
+        auto joined_transitions = transitions->join(ctx_expression_tables->m_transitions);
```

Comment:
Can do this `join` once per context?  The column sets do not change per `gnode.process()` call.